src/test/sql/sql_translator_test.cpp

```diff
@@ -640,4 +640,85 @@ TEST_F(SQLTranslatorTest, ColumnAlias) {
+        PredicateNode::make(_table_a_b, PredicateCondition::NotLike, std::string("special request"),
+          _stored_table_node_a),
+        _stored_table_node_b));
+  // clang-format on
+
+  EXPECT_LQP_EQ(expected, result_node);
+}
+
+TEST_F(SQLTranslatorTest, MultipleJoinConditionsOnRightSide) {
+  const auto query =
+      "SELECT * FROM table_a JOIN table_b ON table_a.a = table_b.a AND table_b.b NOT LIKE 'special request'";
+  auto result_node = compile_query(query);
+
+  // clang-format off
+  const auto expected = ProjectionNode::make_pass_through(
+    JoinNode::make(JoinMode::Inner, LQPColumnReferencePair{_table_a_a, _table_b_a}, PredicateCondition::Equals,
+      _stored_table_node_a,
+      PredicateNode::make(_table_b_b, PredicateCondition::NotLike, std::string("special request"),
+        _stored_table_node_b)));
+  // clang-format on
+
+  EXPECT_LQP_EQ(expected, result_node);
+}
+
+TEST_F(SQLTranslatorTest, MultipleJoinConditionsOnBothSides) {
+  const auto query =
+      "SELECT * FROM table_a JOIN table_b ON table_a.a = table_b.a "
+      "AND table_a.b != 3 "
+      "AND table_a.a == table_a.b "
+      "AND table_b.b NOT LIKE 'special request' "
+      "AND table_b.b == table_b.a ";
+  auto result_node = compile_query(query);
+
+  // clang-format off
+  const auto expected = ProjectionNode::make_pass_through(
+    JoinNode::make(JoinMode::Inner, LQPColumnReferencePair{_table_a_a, _table_b_a}, PredicateCondition::Equals,
+
+      PredicateNode::make(_table_a_a, PredicateCondition::Equals, _table_a_b,
+        PredicateNode::make(_table_a_b, PredicateCondition::NotEquals, static_cast<int64_t>(3),
+          _stored_table_node_a)),
+
+      PredicateNode::make(_table_b_b, PredicateCondition::Equals, _table_b_a,
+        PredicateNode::make(_table_b_b, PredicateCondition::NotLike, std::string("special request"),
+          _stored_table_node_b))));
+  // clang-format on
+
+  EXPECT_LQP_EQ(expected, result_node);
+}
+
+TEST_F(SQLTranslatorTest, JoinConditionAmbiguous) {
+  const auto query = "SELECT * FROM table_a JOIN table_b ON a = b";
+  EXPECT_THROW(compile_query(query), std::runtime_error);
+}
+
+TEST_F(SQLTranslatorTest, NonJoinConditionAmbiguous) {
+  const auto query = "SELECT * FROM table_a JOIN table_b ON table_a.a = table_b.a AND a = 3";
+  EXPECT_THROW(compile_query(query), std::runtime_error);
+}
+
+TEST_F(SQLTranslatorTest, ColumnsOfJoinConditionPermuted) {
+  const auto query = "SELECT * FROM table_a JOIN table_b ON table_b.a = table_a.b";
+  auto result_node = compile_query(query);
+
+  // clang-format off
+  const auto expected = ProjectionNode::make_pass_through(
+    JoinNode::make(JoinMode::Inner, LQPColumnReferencePair {_table_a_b, _table_b_a}, PredicateCondition::Equals,
+      _stored_table_node_a,
+      _stored_table_node_b));
+  // clang-format on
+
+  EXPECT_LQP_EQ(expected, result_node);
+}
+
```

Comment:
I just realized that all tests deal with inner joins. Please remind me what the status was on outer joins.

Reply:
I wrote these tests because I needed to cover the code I was going to refactor in fdab01c.

But that aside, we settled for a general purpose solution which generates valid LQPs independently of the join type. That means, these tests do cover the code paths regardless of the join type.
Key point for outer joins is that predicates for the null-supplying side must not appear after the join-node in the LQP. But since the procedure for any of the join types is exactly the same and we are always inserting the non-join predicates before the join node, I think parameterising the test with the join type and to expect the same outcome in all cases would be overdoing it.

Reply:
Could you please point me to the line that makes sure that the pushdown only happens for predicates of inner joins? I seem to be blind here. :/

Reply:
You are possibly using the same mental model for translating join conditions and the predicate pushdown rule (PPR) . Albeit doing similar things wrt. predicates and subplans, different constraints may or may not apply.

The PPR is currently limited to inner joins for simplicity. It does not incorporate knowledge about outer joins and their null-supplying sides etc. The join type 'inner' does not enforce additional constraints on the validity of the pushdown.

Translating join conditions on the other hand is applicable to all join types. This is due to the fact that given a join and its associated conditions, the join is executed as if the predicates exclusive to one input side were applied beforehand and then the matchmaking took place.

I currently cannot think of a reason for restricting the join condition problem to inner joins.

Reply:
http://sqlfiddle.com/#!9/fd8fe9/5

We have to be careful with what predicates we place under the join. No matter if the SQLTranslator puts them there or the pushdown rule.

Inner joins are obviously fine. Placing it in the null-producing side (here, t2) appears to be fine as well. Placing it in the preserving side (t1) is not.

Other than that, I think this PR is ready to go.